src/source/channel_router.rs

```diff
@@ -0,0 +1,306 @@
+    /// to `channel_mappings` into its output samples.
+    ///
+    /// # Panics
+    ///
+    /// - if `channel_count` is not equal to `channel_map`'s second dimension
+    /// - if `input.channels()` is not equal to `channel_map`'s first dimension
+    pub fn new(
+        input: I,
+        channel_count: u16,
+        channel_map: ChannelMap,
+    ) -> (ChannelRouterController, Self) {
+        assert!(channel_count as usize == channel_map[0].len());
+        assert!(input.channels() as usize == channel_map.len());
+
+        let (tx, rx) = channel();
+
+        let controller = ChannelRouterController { sender: tx };
+        let source = Self {
+            input,
+            channel_map,
+            current_channel: channel_count,
+            // this will cause the input buffer to fill on first call to next()
+            channel_count,
+            // channel_count is redundant, it's implicit in the channel_map dimensions
+            // but maybe it's saving us some time, we do check this value a lot.
+            input_buffer: vec![],
+            receiver: rx,
+        };
+
+        (controller, source)
+    }
+
+    /// Destroys this router and returns the underlying source.
+    #[inline]
+    pub fn into_inner(self) -> I {
+        self.input
+    }
+
+    /// Get mutable access to the inner source.
+    #[inline]
+    pub fn inner_mut(&mut self) -> &mut I {
+        &mut self.input
+    }
+}
+
+impl<I> Source for ChannelRouterSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    #[inline]
+    fn current_frame_len(&self) -> Option<usize> {
+        self.input.current_frame_len()
+    }
+
+    #[inline]
+    fn channels(&self) -> u16 {
+        self.channel_count
+    }
+
+    #[inline]
+    fn sample_rate(&self) -> u32 {
+        self.input.sample_rate()
+    }
+
+    #[inline]
+    fn total_duration(&self) -> Option<std::time::Duration> {
+        self.input.total_duration()
+    }
+}
+
+impl<I> Iterator for ChannelRouterSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_channel >= self.channel_count {
+            // We've reached the end of the frame, time to grab another one from the input
+            let input_channels = self.input.channels() as usize;
+
+            // This might be too fussy, a source should never break a frame in the middle of an
+            // audio frame.
+            let samples_to_take = min(
+                input_channels,
+                self.input.current_frame_len().unwrap_or(usize::MAX),
+            );
+
+            // fill the input buffer. If the input is exhausted and returning None this will make
+            // the input buffer zero length
+            self.input_buffer = self.inner_mut().take(samples_to_take).collect();
+
+            self.current_channel = 0;
+
+            for change in self.receiver.try_iter() {
```

Comment:
Yes, it may do as first version, but polling channel is relatively expensive. In other places an additional flag is used to skip the loop altogether when there are no changes (e.g. `mixer::Mixer::has_pending`). See also https://github.com/RustAudio/rodio/issues/658.